ccan/ccan/graphql/graphql.c

```diff
@@ -0,0 +1,1031 @@
+						c = *--p;
+					}
+					str_end = p - 1;
+					if (c != '"')
+						return "unterminated string or invalid character";
+					c = *p++;
+					if (c != '"')
+						return "invalid string termination";
+					c = *p++;
+					if (c != '"')
+						return "invalid string termination";
+				} else {
+					// empty string
+					str_end = str_begin;
+					--p;
+				}
+			} else {
+				// normal string
+				--p;
+				do {
+					c = *p++;
+					if (c == '\\') {
+						c = *p++;
+						if (strchr("\"\\/bfnrtu", c)) {
+							if (c == 'u') {
+								c = *p++;
+								if (!HEX_DIGIT(c))
+									return "invalid unicode escape sequence";
+								c = *p++;
+								if (!HEX_DIGIT(c))
+									return "invalid unicode escape sequence";
+								c = *p++;
+								if (!HEX_DIGIT(c))
+									return "invalid unicode escape sequence";
+								c = *p++;
+								if (!HEX_DIGIT(c))
+									return "invalid unicode escape sequence";
+							} else {
+								c = 'a'; // anything besides a quote to let the loop continue
+							}
+						} else {
+							return "invalid string escape sequence";
+						}
+					}
+				} while (STRING_CHAR(c));
+				if (c != '"')
+					return "unterminated string or invalid character";
+				str_end = p - 1;
+			}
+			int str_len = str_end - str_begin;
+
+			tok = tal(tok_list, struct graphql_token);
+			list_add_tail(tok_list, &tok->list);
+			tok->token_type = 's';
+			tok->token_specific = 's';
+			tok->token_string = tal_strndup(tok, str_begin, str_len);
+			tok->source_line = line_num;
+			tok->source_column = str_begin - line_beginning + 1;
+			tok->source_len = str_len;
+
+			// Process escape sequences. These always shorten the string (so the memory allocation is always enough).
+			char d;
+			char *q = tok->token_string;
+			char *rewrite_dest;
+			int quotes = 0;
+			while ((d = *q++)) {
+				if (str_block) {
+					if (d == '\"') quotes++; else quotes = 0;
+					if (quotes == 3 && *(q-4) == '\\') {
+						quotes = 0;
+						rewrite_dest = q - 4;
+						cpystr(rewrite_dest, q - 3);
+					}
+				} else {
+					if (d == '\\') {
+						rewrite_dest = q - 1;
+						d = *q++;
+						switch (d) {
+						case '\"':
+							*rewrite_dest++ = '\"';
+							cpystr(rewrite_dest, q--);
+							break;
+						case 'b':
+							*rewrite_dest++ = '\b';
+							cpystr(rewrite_dest, q--);
+							break;
+						case 'f':
+							*rewrite_dest++ = '\f';
+							cpystr(rewrite_dest, q--);
+							break;
+						case 'n':
+							*rewrite_dest++ = '\n';
+							cpystr(rewrite_dest, q--);
+							break;
+						case 'r':
+							*rewrite_dest++ = '\r';
+							cpystr(rewrite_dest, q--);
+							break;
+						case 't':
+							*rewrite_dest++ = '\t';
+							cpystr(rewrite_dest, q--);
+							break;
+						case 'u': {
+								// Insert escaped character using UTF-8 multi-byte encoding.
+								char buf[] = {*q++, *q++, *q++, *q++, 0};
+								int code_point = strtol(buf, 0, 16);
+								int bytes = utf8_encode(code_point, rewrite_dest);
```

Comment:
This can fail!  And return 0.

Reply:
I added a comment to indicate that this was taken into consideration and that a zero size will cleanly result in omitting the offending character.